metadata/reportssql/index_program_functions.sql

```diff
@@ -207,27 +273,129 @@ BEGIN
+BEGIN
+    DECLARE result TINYINT(1) DEFAULT 0;
+
+    SELECT TRUE INTO result
+    FROM relationship r
+        JOIN person pIndex ON (r.person_a = p_patientId AND r.person_b = pIndex.person_id) OR
+            (r.person_a = pIndex.person_id AND r.person_b = p_patientId)
+        JOIN relationship_type rt ON r.relationship = rt.relationship_type_id  AND retired = 0
+    WHERE r.voided = 0 AND
+        getPatientIndexTestingDateAccepted(pIndex.person_id) IS NOT NULL AND
+        rt.a_is_to_b = "RELATIONSHIP_PARTNER"
+    LIMIT 1;
+
+    RETURN result;
+END$$
+DELIMITER ;
+
+-- patientWithIndexChild
+
+DROP FUNCTION IF EXISTS patientWithIndexChild;
+
+DELIMITER $$
+CREATE FUNCTION patientWithIndexChild(
+    p_parent INT(11)) RETURNS TINYINT(1)
+    DETERMINISTIC
+BEGIN
+    DECLARE result TINYINT(1) DEFAULT 0;
+
+    SELECT TRUE INTO result
+    FROM relationship r
+        JOIN person pIndexChild ON r.person_a = pIndexChild.person_id AND r.person_b = p_parent
+        JOIN relationship_type rt ON r.relationship = rt.relationship_type_id  AND retired = 0
+    WHERE r.voided = 0 AND
+        getPatientIndexTestingDateAccepted(pIndexChild.person_id) IS NOT NULL AND
+        (rt.b_is_to_a = "RELATIONSHIP_BIO_CHILD")
     LIMIT 1;
 
     RETURN result;
 END$$
 DELIMITER ;
+
+-- patientWithIndexParent
+
+DROP FUNCTION IF EXISTS patientWithIndexParent;
+
+DELIMITER $$
+CREATE FUNCTION patientWithIndexParent(
+    p_child INT(11)) RETURNS TINYINT(1)
+    DETERMINISTIC
+BEGIN
+    DECLARE result TINYINT(1) DEFAULT 0;
+
+    SELECT TRUE INTO result
+    FROM relationship r
+        JOIN person pIndexParent ON r.person_b = pIndexParent.person_id AND r.person_a = p_child
+        JOIN relationship_type rt ON r.relationship = rt.relationship_type_id  AND retired = 0
+    WHERE r.voided = 0 AND
+        getPatientIndexTestingDateAccepted(pIndexParent.person_id) IS NOT NULL AND
+        (rt.b_is_to_a = "RELATIONSHIP_BIO_CHILD")
+    LIMIT 1;
+
+    RETURN result;
+END$$
+DELIMITER ;
+
+-- getPatientIndexTestingDateOffered
+
+DROP FUNCTION IF EXISTS getPatientIndexTestingDateOffered;
+
+DELIMITER $$
+CREATE FUNCTION getPatientIndexTestingDateOffered(
+    p_patientId INT(11)) RETURNS DATE
+    DETERMINISTIC
+BEGIN
+    DECLARE uuidIndexTestingDateOffered VARCHAR(38) DEFAULT "836fe9d4-96f1-4fea-9ad8-35bd06e0ee05";
+    RETURN getObsDatetimeValue(p_patientId, uuidIndexTestingDateOffered);
+END$$
+DELIMITER ;
+
+-- getIndexType
+
+DROP FUNCTION IF EXISTS getIndexType;
+
+DELIMITER $$
+CREATE FUNCTION getIndexType(
+    p_patientId INT(11)) RETURNS VARCHAR(255)
+    DETERMINISTIC
+BEGIN
+    DECLARE daysBetweenHIVPosAndART INT(11) DEFAULT getDaysBetweenHIVPosAndART(p_patientId);
+    DECLARE viralLoadResult INT(11) DEFAULT getViralLoadTestResult(p_patientId);
+
+    IF (daysBetweenHIVPosAndART > 0 AND daysBetweenHIVPosAndART <= 30) THEN
+        RETURN 'Index Case New HTS POS and initiated on treatment within a month';
+    ELSEIF (viralLoadResult > 1000) THEN
```

Comment:
@lauravignoli the above condition looks different to the rest of the conditions (that are all related to daysBetweenHIVPosAndART). Isn't it possible to have a case where a patient is both 'Index Case virally unsuppressed client' and 'Index Case New HTS POS and initiated on treatment within a month'? Could we check the logic with Emeka and/or JP?

Reply:
@tmvumbi2 I checked with Emeka before writing this:

![indexType](https://user-images.githubusercontent.com/1198827/83852037-7b1df700-a713-11ea-86e1-e5a9177e62df.png)